File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        l, r = 0, 0
        maxLen = -1
        hs = set()
        while (r < len(s)):
            if s[r] not in hs:
                hs.add(s[r])
                r += 1
                maxLen = max(maxLen, r - l)
            else:
                hs.remove(s[l])
                l += 1
        return maxLen

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLen = 0;
        Set<Character> hs = new HashSet<>();
        int l = 0;

        for (int r = 0; r < n; r++){
            if (!hs.contains(s.charAt(r))){
                hs.add(s.charAt(r));
                maxLen = Math.max(maxLen, r - l + 1);
            }else{
                while (hs.contains(s.charAt(r))){
                    hs.remove(s.charAt(l));
                    l++;
                }
                hs.add(s.charAt(r));
            }
        }
        return maxLen;
    }
}